### ISSUE TITLE
Drop IGatewayListProvider

### ISSUE THREAD

**Issue by galvesribeiro** [dotnet/orleans#4006]
Currently in order to `IClusterClient` to communicate with the silos, it has to have access to the list of gateways available in the cluster by using the `IGatewayListProvider`. 

`IGatewayListProvider` implementations in other hand, has the solo responsibility thru `IGatewayListProvider.GetGateways()` to connect to the membership storage, read all the silo records and filter out silos where status is `Dead` or that doesn't have a gateway installed (`ProxyPort` != 0). It also periodically refresh the gateway list in a (configurable) time interval.

Orleans was designed to have silos in a trusted network, the clients in other hand, usually are exposed to a non-trusted network or DMZ where it receive requests (i.e. public internet). 

With that in mind, having the client to have access to the membership storage today seems something not really ideal. You not just have to set credentials so the client can connect to it, but you open possibility to "rogue" clients to even modify the cluster storage which can led to serious problems specially if the developer is hosting cluster + state + reminders storage in a single backend storage.

One alternative to protect against that, is to create a different set of credentials with limited access to the cluster storage but maintain that can be very cubersome...

I would like to discuss the possibility of drop `IGatewayListProvider` and have Orleans client to connect to any gateway at startup and send an initialization message that request for the list of gateways available. The logic would be exactly the same as Today for refresh the gateway list as well, but it would instead of read from the datastorage, send a message to the cluster in whatever gateway available and have the list of the gateways returned. That would remove a need to maintain those provider and let the user choose the best option for their use case to expose the cluster gateways to the network (firewall, load balancer, services, etc.).

What do you guys think?

Thanks

**Comment by ReubenBond** [comment]
`IGatewayListProvider` doesn't necessarily require hitting the same backend store as `IMembershipTable`. It could be implemented based on DNS or using any other service discovery mechanism. You could also implement it as you say, by connecting to some node on startup and discovering gateways using those nodes. Of course, you still have to have some way to find those nodes (eg, DNS) in the first place.

If it were dropped, what would the replacement look like?

**Comment by galvesribeiro** [comment]
Those are some of the problematic scenarios that can be covered:

1. IT don't want to expose all silos (even if _just_ the gateway ones) thru the DMZ Firewall, so they have a single IP exposed which acts as a NLB and only packets from that particular IP is allowed to enter the trusted Zone. That would make use of something like the `UseStaticGatewayListProvider` with a single IP. It reduces the need for IT to manage firewall rules for every silo specially if we are talking about containers that comes and goes with different IP addresses.
2. Lets say IT is willing to expose every silo IP to the DMZ, the client still need a way to enumerate them and Today, you must either know before hand the list of IPs (which is impossible with container scenario as mentioned on 1) and use `UseStaticGatewayListProvider` or, use a membership provider (`IGatewayListProvider`) which today requires (by default, without custom code) to connect to some sort of source of truth, the backend storage or service discovery mechanism, in order to figure out the gateway list. That requires not just the client to "know" the type of provider but also provide eventual credentials to it.
3. Having one `IGatewayListProvider` per backend makes the client sort of non-portable as we require to change the package/implementation of the provider as the backend changes.

We may keep the gateway provider and use it if registered, but what I'm suggesting is that we have a way for the client to:

1. Connect to the cluster by sending a specific `GatewayDiscovery` message to any of the gateways if it knows at least 1 IP before hand.
2. If it doesn't know the fixed IP of a gateway (mostly likely), send a  `GatewayDiscovery` message to pre-defined fixed service IP, which could be exposed by the IT as a load balancer and would route to any gateway behind it.
3. Allow the client to ignore a "list" of gateways and only send all messages thru a single IP defined by IT which would be a load balancer and forward to a random gateway.

That way you would: 

1. Avoid people to use/implement/configure a provider
2. Portability of code across multiple clusters
3. Streamline connections from client to gateways in a uniform way
4. Don't require credentials for membership storage
5. Allow more granular control to IT without compromise features for the developer
6. Reduce the complexity of silos exposing real/public vs listening/internal proxy port as we see on scenarios like containers and are generating so much confusion

When I say IP, it can be a DNS name. I'm more referring to an endpoint resolvable at runtime.

**Comment by galvesribeiro** [comment]
Scenarios like https://github.com/OrleansContrib/Orleans.Clustering.Kubernetes/issues/6 would be unnecessary to be implemented if we had what I'm suggesting here.

**Comment by ReubenBond** [comment]
Why would you expose *any* silos through DMZ or *any* ports on any silos? The silo must not be exposed at all, gateway port or otherwise.

I'm still not seeing how you couldn't implement an `IGatewayListProvider` which worked the way you're describing. The abstraction is very general, it could work with any service discovery mechanism, including ones which involve bootstrapping from a static set of addresses (or a single address, or some external mechanism like DNS).

**Comment by galvesribeiro** [comment]
> Why would you expose any silos through DMZ or any ports on any silos? The silo must not be exposed at all, gateway port or otherwise.

In order to a client on DMZ to connect to the silos on the private/trusted network, we need to reach the silos with the Gateway port open, hence why we need to open the Gateway port at the firewall between DMZ and the private network.

> I'm still not seeing how you couldn't implement an `IGatewayListProvider` which worked the way you're describing.

Like I said on the previous message, we may keep the abstraction.

Let me picture the scenarios:

![image](https://user-images.githubusercontent.com/4714040/36073644-17e6d126-0f1b-11e8-9ad2-1eff273143b0.png)

In this scenario, the Client knows at startup the IP/port list of all gateway, so it is fine to use the logic implemented on `UseStaticGatewayListProvider` as the user intent to only use that static list.

![image](https://user-images.githubusercontent.com/4714040/36073669-7c95ca64-0f1b-11e8-92b8-eb140421356f.png)

In that scenario however, IT team exposes a single IP and allow only that IP to be accessible thru the DMZ firewall. The client has only 1 gateway, always. Even that behind the scenes the client will be routed the connection to different gateways it still _think_ it is on a single gateway.

![image](https://user-images.githubusercontent.com/4714040/36073749-f7634df6-0f1c-11e8-9c73-acb8b9a7cdc7.png)

And in the last case, IT exposed both the NLB IP and the individual gateway IPS. The client still only know the NLB IP, so it will send a message to the NLB which will route to any of the gateways available. That message is just saying "Hey, I'm on ClusterId X. Gimme the gateways available". The gateway would return a list of gateways available and the client can then pick one and connect using the same logic as for today.

The point is, on none of those cases we required code or credentials to access a backend storage. All you need is 1 IP for either query for the list or to act as a single endpoint, or a list of fixed gateways.

**Comment by SebastianStehle** [comment]
Short question: Why is your client in DMZ? Usually you have loadbalancer in front of your clients.

**Comment by galvesribeiro** [comment]
@SebastianStehle the balancer for the clients is in front of the DMZ.

**Comment by jms69** [comment]
@ReubenBond  Why can't the gateway be exposed?  If the gateway was locked down via some Mutual Auth mechanism, why do you say the gateway can't be exposed?

I would like to have my silos deployed as a self-contained unit say in Azure Cloud Service or VM scaleset, then have other components like a web app deployed as Application Service talk to it through the gateway.  No Vnet and no DMZ.

**Comment by ReubenBond** [comment]
@jms69 that's just not the way it's designed today.

Today, silos *must* not be exposed to the Internet.
Clients *must* be able to directly address silos.

**Comment by ReubenBond** [comment]
Having a load balancer in front of the gateways breaks the current protocol and reduces our ability to optimize routing in future (eg, direct routing to the correct silo to improve latency and reduce load).

So I'm not in favor of that route in the general case.

At some point, 'dumb' clients are useful to have, eg for an HTTP facade, but I believe we should not break the assumptions of the existing TCP protocol.

**Comment by galvesribeiro** [comment]
Ok, but with the scenarios I've shown, only one (use a single IP as a gateway) _breaks_ anything... All others does the same thing as the current provider model, but don't rely on storage nor custom providers for that.

Also, by dropping the single IP capability, we are closing the doors for multiple corporate environments that separate DMZ from private network as they won't allow the gateways to be opened one by one. Specially if they have a dynamic environment with machines (gateways) up and down on demand...

**Comment by ReubenBond** [comment]
Clients should be on the same network as Silos. If they want to communicate with clients via HTTPS, then that's fine, they can do that and put an LB in front of clients.

I think of Orleans clients as another node type in the same cluster.

In the 2.1 time frame, we'll likely have 'local client' and hopefully HTTP (in my head, though, HTTP will be in 2.2, since I want to focus on perf in 2.1), so the distinction will blur.

**Comment by galvesribeiro** [comment]
I see... So we would have to have an extra layer of services... For example in a e-commerce site I will need:

1. Frontend balancer/firewall
2. Web frontend (DMZ)
3. Web API (DMZ)
4. Internal balancer/firewall
5. Web API (Orleans Client/private network)
6. Orleans Silos (private network)

That looks to me a (regardless of the existence of 'local client') to be too much overhead. The gateway is supposed to be... a gateway. The front door of the cluster to the client. The only client I see in this scenario is the (3) Web API. 

Somewhere in the future, the Gateway should even be the place where (if we ever have) some authz code would live.

Otherwise, the Gateway functionality doesn't look much useful  to me and could be accomplished by a regular silo.

**Comment by ReubenBond** [comment]
Why would Web frontend be in front of Web API? they should be side-by-side, otherwise is the Web frontend forwarding calls to the Web API?

With local client, that becomes:
1. Frontend LB/FW
2. Frontend HTTP (DMZ, AuthN)
3. Internal LB/FW
4. Silo with LocalClient (exposing your specific HTTP API, AuthZ)

**Comment by ReubenBond** [comment]
Orleans clients today aren't thin - they have some knowledge of the cluster, eg the mapping from interfaces to grain implementations. If you want increased security via an internal FW, I would recommend having a well-defined API for communication across that FW. I.e, don't use `IClusterClient` (which lets you interact with the cluster however you want), use a well-defined HTTPS API which only exposes the necessary functionality.

**Comment by galvesribeiro** [comment]
> Why would Web frontend be in front of Web API

Not in front, just in the same DMZ network. Just didn't found a way to type properly on bullets 😄 .

Regardless of the client be local or not. The point is, we will have to manage HTTP services in two sides, and that in some scenarios is not desirable.

But anyway, I got your point. I believe this will never change and for me (and others who end up with complex deployments that is more than a simple network with a silo and a webAPI) it is unfortunate and frustrating.

But thanks for expose your ideas.

**Comment by ReubenBond** [comment]
The gateway isn't designed to be a secure entry point to the cluster: clients have the same level of access as silos. If you want to *work around* network security restrictions, then you can by all means expose a single gateway address and use that. It will work, probably forever. Gateways are definitely not designed as some kind of security boundary, though, so doing so really has no benefit security-wise.

What I'm saying is that that's not secure and therefore not a recommended approach.

It might also require some modification of settings/code to ensure that that single address is never internally blacklisted by the client when one of the nodes behind the LB goes down and the connection is dropped temporarily.

The assumption for gateways today is that there can be more than one, but that some silos might not run a gateway while others do. Therefore, any optimization would have to be opportunistic anyway (eg: route to the activation's silo if we know it and can access it, otherwise load balance over gateways).

**Comment by sergeybykov** [comment]
> clients have the same level of access as silos.

I'd like to double-stress that. For example, a client can make calls to system targets, such as `IRemoteGrainDirectory`, `ISiloControl`, etc. So it does not provide any extra security, not even much of obscurity. They main purpose of the gateway is that clients don't have to maintain state on their side about what grains are activated on which silos in the cluster.

A locked down HTTP-based API (ideally, with API management layer and throttling) is usually the right thing to expose to any partially trusted or untrusted environment. That's what the client frontends are generally used for.

**Comment by galvesribeiro** [comment]
OK, thanks for for all the inputs. I've discussed this with Reuben offline and we came to the conclusion that all this will be gone once he had time to finish the `LocalClient`.

With that, hopefully people don't need to create and maintain (for example) API controllers etc. As long as it behave as a new _Endpoint provider_ model, where people just plug the transport, we shouldn't care about client anymore and all the side-effects it brings in complex scenarios.

I'm closing the issue as I'm good with the whole discussion and will be waiting for `LocalClient` to get out (hurry up Reuben! 😄).

Thanks for the input guys!
